### PR TITLE
Trie hash optimizations 

### PR DESCRIPTION
**Pull request by dvush**
This PR improves a performance of the trie when calculating hash.

In my benchmark it reduces time from 100ms to 90ms and this 10ms are more noticeable in implementation that caches db reads for the trie.

The most influential commits are:
* [cursor subnode full_key cache](https://github.com/paradigmxyz/reth/commit/6ba5d99e39cfe1fd04050e090c9f2d7cee87330b)
This key clone alone is mostly responsible for 10ms drop
* [AHash in trie](https://github.com/paradigmxyz/reth/commit/397f236e3f8b6b24ce38cfa04c734cd02a2bc6d1)
Recently hashbrown was removed from the repo and it made all hash maps / sets much slower. Its mostly due to hashbrown using ahash for hashing while std is using siphash. This only changes hashmaps in trie related code but other parts of reth can benefit from it too. 
* [prealloc children](https://github.com/paradigmxyz/reth/commit/bc60f6865c09a67709f05d228e25a3e3ab4018fc) is not that noticeable but improves `push_branch_node` a bit. (without it about 20% of the method is spent doing this collect) (moved to https://github.com/alloy-rs/trie/pull/1)

### CODE REVIEW

**Comment by dvush** [comment]
Depends on https://github.com/DaniPopes/nybbles/pull/1

**Comment by onbjerg** [comment]
cc @DaniPopes this should be in alloy-trie, correct?

**Comment by DaniPopes** [comment]
> cc @DaniPopes this should be in alloy-trie, correct?

Only the primitives change. If you don't mind you can open a PR to https://github.com/alloy-rs/trie with the second commit @dvush

File: crates/trie/src/walker.rs [hunk 1]
```diff
@@ -235,7 +234,7 @@ impl<C: TrieCursor> TrieWalker<C> {
     }
 
     fn update_skip_node(&mut self) {
-        self.can_skip_current_node = if let Some(key) = self.key() {
+        self.can_skip_current_node = if let Some(key) = self.stack.last().map(|n| n.full_key()) {
```
Comment (DaniPopes) [inline crates/trie/src/walker.rs:237]: Is this not the same as `self.key()`?

**Review by DaniPopes — CHANGES_REQUESTED** [review]

File: crates/trie/src/walker.rs (hunk 1 above)
Comment (dvush) [reply crates/trie/src/walker.rs:237]: yes!
unfortunately line below does mutable borrow of self and if its done with key() method borrow checker things that the whole self is borrowed but when inlined this way it knows that we are working with different fields

a bit better look would be to move key method to self.stack, maybe?

**Comment by DaniPopes** [comment]
Released [nybbles 0.1.2](https://docs.rs/nybbles/0.1.2/nybbles/)

File: crates/trie/src/walker.rs (hunk 1 above)
Comment (DaniPopes) [reply crates/trie/src/walker.rs:237]: I see, this makes sense. Can also clean this up:
```diff
--- a/crates/trie/src/walker.rs
+++ b/crates/trie/src/walker.rs
@@ -234,10 +234,8 @@ impl<C: TrieCursor> TrieWalker<C> {
     }
 
     fn update_skip_node(&mut self) {
-        self.can_skip_current_node = if let Some(key) = self.stack.last().map(|n| n.full_key()) {
-            let contains_prefix = self.changes.contains(key);
-            let hash_flag = self.stack.last().unwrap().hash_flag();
-            !contains_prefix && hash_flag
+        self.can_skip_current_node = if let Some(node) = self.stack.last() {
+            node.hash_flag() && !self.changes.contains(node.full_key())
         } else {
             false
         };
```

**Review by DaniPopes — APPROVED** [review]
LGTM after comments are resolved. cc @rkrasiuk

File: crates/trie/src/walker.rs (hunk 1 above)
Comment (dvush) [reply crates/trie/src/walker.rs]: nice, committed your suggestion

**Comment by rkrasiuk** [comment]
@dvush in any follow-up perf PRs it would be great to see side by side comparison of flamegraphs / benches before and after the change. otherwise, it's tough to verify the performance boost from the changes. even some obvious "improvements" might degrade the performance as we've experienced in the past

**Review by rkrasiuk — APPROVED** [review]
overall lgtm modulo my comment re perf PRs above. tested at mainnet tip and ran execution debug script on holesky

**Pull request merged**